core/src/nucleus/ribosome/mod.rs

```diff
@@ -0,0 +1,349 @@
+// HC API FUNCTION IMPLEMENTATIONS
+//--------------------------------------------------------------------------------------------------
+
+/// List of all the API functions available in Nucleus
+#[repr(usize)]
+enum HcApiFuncIndex {
+    /// Error index for unimplemented functions
+    MissingNo = 0,
+    /// Print debug information in the console
+    /// print(s : String)
+    Print,
+    /// Commit an entry to source chain
+    /// commit(entry_type : String, entry_content : String) -> Hash
+    Commit,
+    /// Get an entry from source chain by key (header hash)
+    /// get(key: String) -> Pair
+    Get,
+    // Add new API function index here
+    // ...
+}
+
+//--------------------------------------------------------------------------------------------------
+// Wasm call
+//--------------------------------------------------------------------------------------------------
+
+/// Object holding data to pass around to invoked API functions
+#[derive(Clone, Debug)]
+pub struct Runtime {
+    print_output: String,
+    pub result: String,
+    action_channel: Sender<state::ActionWrapper>,
+    observer_channel: Sender<Observer>,
+    memory_manager: SinglePageManager,
+}
+
+/// take standard, memory managed runtime argument bytes, extract and convert to serialized struct
+pub fn runtime_args_to_utf8(runtime: &Runtime, args: &RuntimeArgs) -> String {
+    // @TODO don't panic in WASM
+    // @see https://github.com/holochain/holochain-rust/issues/159
+    assert_eq!(1, args.len());
+
+    // Read complex argument serialized in memory
+    let encoded_allocation: u32 = args.nth(0);
+    let allocation = SinglePageAllocation::new(encoded_allocation);
+    let allocation = allocation
+        // @TODO don't panic in WASM
+        // @see https://github.com/holochain/holochain-rust/issues/159
+        .expect("received error instead of valid encoded allocation");
+    let bin_arg = runtime.memory_manager.read(allocation);
+
+    // deserialize complex argument
+    String::from_utf8(bin_arg)
+        // @TODO don't panic in WASM
+        // @see https://github.com/holochain/holochain-rust/issues/159
+        .unwrap()
+}
+
+/// given a runtime and a string (e.g. JSON serialized data), allocates bytes and encodes to memory
+/// returns a Result suitable to return directly from a zome API function
+pub fn runtime_allocate_encode_str(
+    runtime: &mut Runtime,
+    s: &str,
+) -> Result<Option<RuntimeValue>, Trap> {
+    // write str to runtime memory
+    let mut s_bytes: Vec<_> = s.to_string().into_bytes();
+    s_bytes.push(0); // Add string terminate character (important)
+
+    let allocation_of_result = runtime.memory_manager.write(&s_bytes);
+    if allocation_of_result.is_err() {
+        return Err(Trap::new(TrapKind::MemoryAccessOutOfBounds));
+    }
+
+    let encoded_allocation = allocation_of_result
+        // @TODO don't panic in WASM
```

Comment:
this can't panic because is_err() is done before

Reply:
yeah there's something similar in the get sensor, but also panicking in wasm is totally undebuggable as far as i can tell, because the wasm machinery drops all the relevant information, so we shouldn't even use it for that